From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: GliczDev <67753196+GliczDev@users.noreply.github.com>
Date: Sun, 1 Sep 2024 16:14:44 +0200
Subject: [PATCH] Player events


diff --git a/net/minecraft/server/network/ServerGamePacketListenerImpl.java b/net/minecraft/server/network/ServerGamePacketListenerImpl.java
index 470aeb9eaa7ec7fd64c36f2610f5d3ea40f85c9f..3baed91b086ddb6e1b3b67e7c33585d361cac6f9 100644
--- a/net/minecraft/server/network/ServerGamePacketListenerImpl.java
+++ b/net/minecraft/server/network/ServerGamePacketListenerImpl.java
@@ -1256,11 +1256,22 @@ public class ServerGamePacketListenerImpl
     private void removePlayerFromWorld() {
         this.chatMessageChain.close();
         this.server.invalidateStatus();
+        // Airflow start - Player events
+        /*
         this.server
             .getPlayerList()
             .broadcastSystemMessage(Component.translatable("multiplayer.player.left", this.player.getDisplayName()).withStyle(ChatFormatting.YELLOW), false);
+         */
+        // Airflow end - Player events
         this.player.disconnect();
-        this.server.getPlayerList().remove(this.player);
+        // Airflow start - Player events
+        Component message = this.server.getPlayerList().remove(this.player);
+        if (message != null) {
+            this.server
+                    .getPlayerList()
+                    .broadcastSystemMessage(message, false);
+        }
+        // Airflow end - Player events
         this.player.getTextFilter().leave();
     }
 
diff --git a/net/minecraft/server/players/PlayerList.java b/net/minecraft/server/players/PlayerList.java
index f0fa78caff92a36b2814ec7465212a7e389d2103..4778b96088ffdf19712a476f34324c54e13bd6de 100644
--- a/net/minecraft/server/players/PlayerList.java
+++ b/net/minecraft/server/players/PlayerList.java
@@ -217,8 +217,8 @@ public abstract class PlayerList {
         } else {
             mutableComponent = Component.translatable("multiplayer.player.joined.renamed", player.getDisplayName(), string);
         }
+        mutableComponent.withStyle(ChatFormatting.YELLOW); // Airflow - Player events
 
-        this.broadcastSystemMessage(mutableComponent.withStyle(ChatFormatting.YELLOW), false);
         serverGamePacketListenerImpl.teleport(player.getX(), player.getY(), player.getZ(), player.getYRot(), player.getXRot());
         ServerStatus status = this.server.getStatus();
         if (status != null && !cookie.transferred()) {
@@ -229,6 +229,14 @@ public abstract class PlayerList {
         this.players.add(player);
         this.playersByUUID.put(player.getUUID(), player);
         this.broadcastAll(ClientboundPlayerInfoUpdatePacket.createPlayerInitializing(List.of(player)));
+        // Airflow start - Player events
+        me.glicz.airflow.api.event.player.PlayerJoinEvent event = this.server.getDedicatedServer().airflow.serverEventBus.dispatchPlayerJoin(player, mutableComponent);
+
+        Component message = me.glicz.airflow.util.MinecraftComponentSerializer.INSTANCE.serializeOrNull(event.getMessage());
+        if (message != null) {
+            this.broadcastSystemMessage(message, false);
+        }
+        // Airflow end - Player events
         this.sendLevelInfo(player, serverLevel);
         serverLevel.addNewPlayer(player);
         this.server.getCustomBossEvents().onPlayerConnect(player);
@@ -321,9 +329,22 @@ public abstract class PlayerList {
         }
     }
 
-    public void remove(ServerPlayer player) {
+    public Component remove(ServerPlayer player) { // Aiflow - Player events
         ServerLevel serverLevel = player.serverLevel();
         player.awardStat(Stats.LEAVE_GAME);
+        // Airflow start - Player events
+        if (player.containerMenu != player.inventoryMenu) {
+            player.closeContainer();
+        }
+
+        me.glicz.airflow.api.event.player.PlayerQuitEvent event = player.server.getDedicatedServer().airflow.serverEventBus.dispatchPlayerQuit(
+                player,
+                Component.translatable("multiplayer.player.left", player.getDisplayName())
+                        .withStyle(ChatFormatting.YELLOW)
+        );
+
+        player.doTick();
+        // Airflow end - Player events
         this.save(player);
         if (player.isPassenger()) {
             Entity rootVehicle = player.getRootVehicle();
@@ -353,6 +374,7 @@ public abstract class PlayerList {
         }
 
         this.broadcastAll(new ClientboundPlayerInfoRemovePacket(List.of(player.getUUID())));
+        return me.glicz.airflow.util.MinecraftComponentSerializer.INSTANCE.serializeOrNull(event.getMessage()); // Airflow - Player events
     }
 
     @Nullable
