From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: GliczDev <67753196+GliczDev@users.noreply.github.com>
Date: Sun, 15 Sep 2024 13:34:31 +0200
Subject: [PATCH] Better terminal


diff --git a/net/minecraft/server/MinecraftServer.java b/net/minecraft/server/MinecraftServer.java
index 27a0456799eee616b2d02655595ecd091c8cb871..ce6f381dbd8fbe0c444fa1d9cbb3ddea9ebda4e5 100644
--- a/net/minecraft/server/MinecraftServer.java
+++ b/net/minecraft/server/MinecraftServer.java
@@ -1213,7 +1213,11 @@ public abstract class MinecraftServer extends ReentrantBlockableEventLoop<TickTa
 
     @Override
     public void sendSystemMessage(Component component) {
-        LOGGER.info(component.getString());
+        // Airflow start - Better terminal
+        LOGGER.info(me.glicz.airflow.util.AdventureUtils.ANSI_SERIALIZER.serialize(
+                me.glicz.airflow.util.MinecraftComponentSerializer.INSTANCE.deserialize(component)
+        ));
+        // Airflow end - Better terminal
     }
 
     public KeyPair getKeyPair() {
diff --git a/net/minecraft/server/dedicated/DedicatedServer.java b/net/minecraft/server/dedicated/DedicatedServer.java
index 77dfb2c9692b618b415cf923c46f668f46aee571..99178ac91db13b4a14189411c21f0385916fe481 100644
--- a/net/minecraft/server/dedicated/DedicatedServer.java
+++ b/net/minecraft/server/dedicated/DedicatedServer.java
@@ -107,6 +107,9 @@ public class DedicatedServer extends MinecraftServer implements ServerInterface
         Thread thread = new Thread("Server console handler") {
             @Override
             public void run() {
+                // Airflow start - Better terminal
+                new me.glicz.airflow.console.AirflowConsole(DedicatedServer.this).start();
+                /*
                 BufferedReader bufferedReader = new BufferedReader(new InputStreamReader(System.in, StandardCharsets.UTF_8));
 
                 String string1;
@@ -117,6 +120,8 @@ public class DedicatedServer extends MinecraftServer implements ServerInterface
                 } catch (IOException var4) {
                     DedicatedServer.LOGGER.error("Exception handling console input", (Throwable)var4);
                 }
+                 */
+                // Airflow end - Better terminal
             }
         };
         thread.setDaemon(true);
