From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Pancakse <84547061+Pan-cakse@users.noreply.github.com>
Date: Fri, 25 Oct 2024 04:38:47 +0000
Subject: [PATCH] Rebrand


diff --git a/net/minecraft/server/MinecraftServer.java b/net/minecraft/server/MinecraftServer.java
index 37dba85775f77911af4b242ab8e7b41f4c08db4e..35d5ce35d9cbaf9d3dbcd7ddd8bc023489a25515 100644
--- a/net/minecraft/server/MinecraftServer.java
+++ b/net/minecraft/server/MinecraftServer.java
@@ -1177,7 +1177,7 @@ public abstract class MinecraftServer extends ReentrantBlockableEventLoop<TickTa
 
     @DontObfuscate
     public String getServerModName() {
-        return "vanilla";
+        return "IceCream :)";
     }
 
     public SystemReport fillSystemReport(SystemReport systemReport) {
