From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Panckase <pancakse@icecreammc.xyz>
Date: Sat, 2 Nov 2024 15:24:47 -0600
Subject: [PATCH] Rebrand Server GUI


diff --git a/net/minecraft/server/gui/MinecraftServerGui.java b/net/minecraft/server/gui/MinecraftServerGui.java
index 2da10418a1328773bb7a52da1347fbc8819533d0..b7860c82bb00612e4766bb80597dd8115d122250 100644
--- a/net/minecraft/server/gui/MinecraftServerGui.java
+++ b/net/minecraft/server/gui/MinecraftServerGui.java
@@ -33,8 +33,8 @@ import org.slf4j.Logger;
 public class MinecraftServerGui extends JComponent {
     private static final Font MONOSPACED = new Font("Monospaced", 0, 12);
     private static final Logger LOGGER = LogUtils.getLogger();
-    private static final String TITLE = "Minecraft server";
-    private static final String SHUTDOWN_TITLE = "Minecraft server - shutting down!";
+    private static final String TITLE = "IceCream server";
+    private static final String SHUTDOWN_TITLE = "IceCream server - shutting down!";
     private final DedicatedServer server;
     private Thread logAppenderThread;
     private final Collection<Runnable> finalizers = Lists.newArrayList();
@@ -46,7 +46,7 @@ public class MinecraftServerGui extends JComponent {
         } catch (Exception var3) {
         }
 
-        final JFrame jFrame = new JFrame("Minecraft server");
+        final JFrame jFrame = new JFrame("IceCream server");
         final MinecraftServerGui minecraftServerGui = new MinecraftServerGui(server);
         jFrame.setDefaultCloseOperation(2);
         jFrame.add(minecraftServerGui);
@@ -57,7 +57,7 @@ public class MinecraftServerGui extends JComponent {
             @Override
             public void windowClosing(WindowEvent event) {
                 if (!minecraftServerGui.isClosing.getAndSet(true)) {
-                    jFrame.setTitle("Minecraft server - shutting down!");
+                    jFrame.setTitle("IceCream server - shutting down!");
                     server.halt(true);
                     minecraftServerGui.runFinalizers();
                 }
