From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: GliczDev <67753196+GliczDev@users.noreply.github.com>
Date: Fri, 30 Aug 2024 20:51:42 +0200
Subject: [PATCH] World API


diff --git a/net/minecraft/server/level/ServerLevel.java b/net/minecraft/server/level/ServerLevel.java
index 44e9f311549ae54154941dbe7ad93e6992903ecc..582329cdf456e315cfda0f98d067c6c2fce88523 100644
--- a/net/minecraft/server/level/ServerLevel.java
+++ b/net/minecraft/server/level/ServerLevel.java
@@ -204,6 +204,7 @@ public class ServerLevel extends Level implements WorldGenLevel {
     private final StructureCheck structureCheck;
     private final boolean tickTime;
     private final RandomSequences randomSequences;
+    public final me.glicz.airflow.world.AirWorld airWorld = new me.glicz.airflow.world.AirWorld(this); // Airflow - World API
 
     public ServerLevel(
         MinecraftServer server,
